### PR TITLE
Use custom converter in dateadd function

### PR DESCRIPTION
**Pull request by cbornet**

### CODE REVIEW

File: tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java [hunk 1]
```diff
@@ -105,6 +107,70 @@ public void testPrimitive() throws Exception {
     assertEquals(value.getNativeObject(), "42");
   }
 
+  @Test
+  public void test1() throws Exception {
```
Comment (aymkhalil) [inline tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java:111]: CI is failing on this.

File: pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java [hunk 2]
```diff
@@ -67,33 +66,13 @@ public static long now() {
...
-    Instant instant = Instant.ofEpochMilli(epochMillis);
-    OffsetDateTime localDateTime = OffsetDateTime.ofInstant(instant, ZoneOffset.UTC);
-    return dateadd(localDateTime, delta, unit);
-  }
-
-  private static long dateadd(OffsetDateTime offsetDateTime, long delta, String unit) {
-    final ChronoUnit chronoUnit;
-    switch (unit) {
+    ChronoUnit chronoUnit;
+    switch (CustomTypeConverter.INSTANCE.coerceToString(unit)) {
```
Comment (aymkhalil) [inline pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java:75]: this way, `unit` is more permissive than expected for this use case. what's wrong with limiting it `String` type in the method signature and just using it as is?

**Review by aymkhalil — CHANGES_REQUESTED** [review]
Mainly CI is failing, also was not sure what use case required our custom string coersion for the `unit` part

File: pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java (hunk 2 above)
Comment (cbornet) [reply pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java:79]: It's not really more permissive, the values allowed are still the same. It just allows to use a type that is compatible.
Practically, only BYTES can match.

File: tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java (hunk 1 above)
Comment (cbornet) [reply tests/src/test/java/com/datastax/oss/pulsar/functions/transforms/tests/AbstractDockerTest.java:111]: Oups. This should not have been committed.

File: pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java [hunk 3]
```diff
@@ -119,13 +98,25 @@ private static long dateadd(OffsetDateTime offsetDateTime, long delta, String un
       case "millis":
         chronoUnit = ChronoUnit.MILLIS;
         break;
+      case "nanos":
```
Comment (cbornet) [inline pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java:101]: Added nanos to the party.

File: pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java [hunk 4]
```diff
@@ -119,13 +98,25 @@ private static long dateadd(OffsetDateTime offsetDateTime, long delta, String un
...
+    if (chronoUnit == ChronoUnit.MONTHS || chronoUnit == ChronoUnit.YEARS) {
+      return CustomTypeConverter.INSTANCE
+          .coerceToOffsetDateTime(input)
+          .plus(CustomTypeConverter.INSTANCE.coerceToLong(delta), chronoUnit)
+          .toInstant()
+          .toEpochMilli();
+    }
+    return CustomTypeConverter.INSTANCE
+        .coerceToInstant(input)
+        .plus(CustomTypeConverter.INSTANCE.coerceToLong(delta), chronoUnit)
```
Comment (cbornet) [inline pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java:119]: `plus` operation is faster on Instant.

**Comment by cbornet** [comment]
The unit conversion is because the string could come from a BYTES SchemaType.

**Pull request merged**